### PR TITLE
[FLINK-32032] Upgrade to flink-shaded 17.0

### PR DESCRIPTION
**Pull request by MartijnVisser**
## What is the purpose of the change

* Update to flink-shaded 17.0

## Brief change log

* Updated POM files
* Changed all `guava30` occurrences to `guava31` since flink-shaded uses a newer Guava version
* Updated Curator to 5.4.0

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by MartijnVisser** [comment]
@flinkbot run azure

File: pom.xml [hunk 1]
```diff
@@ -376,7 +376,7 @@ under the License.
 			<dependency>
 				<groupId>org.apache.flink</groupId>
 				<artifactId>flink-shaded-netty</artifactId>
-				<version>4.1.90.Final-${flink.shaded.version}</version>
```
Comment (XComp) [inline pom.xml]: The commit message doesn't match the change? :thinking:

File: pom.xml [hunk 2]
```diff
@@ -122,8 +122,8 @@ under the License.
 		<flink.forkCountUnitTest>4</flink.forkCountUnitTest>
 		<flink.reuseForks>true</flink.reuseForks>
 		<flink.surefire.baseArgLine>-XX:+UseG1GC -Xms256m</flink.surefire.baseArgLine>
-		<flink.shaded.version>16.1</flink.shaded.version>
-		<flink.shaded.jackson.version>2.13.4</flink.shaded.jackson.version>
```
Comment (XComp) [inline pom.xml:126]: There's also `<curator.version/>` property further down in the file that is used for test code. I suggest that we keep these two dependencies in sync. We would have to change some parameter initialization (see [CURATOR-663](https://issues.apache.org/jira/browse/CURATOR-663)) if we decide to upgrade.

**Review by XComp — CHANGES_REQUESTED** [review]
Thanks @MartijnVisser for picking this up. I left two comments. PTAL

File: pom.xml (hunk 2 above)
Comment (MartijnVisser) [reply pom.xml:126]: > I suggest that we keep these two dependencies in sync.

Just to double check, so upgrade to Curator 4.0 as part of the upgrade to flink-shaded 17.0?
If so, can you help with the necessary changes?

File: pom.xml (hunk 1 above)
Comment (MartijnVisser) [reply pom.xml]: I'll change that when I'm rebasing

**Comment by MartijnVisser** [comment]
@XComp Thnx for the review. I've decided to split of the upgrade of Pulsar in flink-python from this PR, because it requires more changes. I've filed https://issues.apache.org/jira/browse/FLINK-32056 for that. When that's done and merged, I'll rebase the PR. If you can help with the necessary Curator changes, we can work a bit in parallel I think

File: pom.xml (hunk 2 above)
Comment (MartijnVisser) [reply pom.xml:126]: Ah yes, that's what I meant but it doesn't help if I type `4.0` instead of `5.4.0` 👍

**Comment by MartijnVisser** [comment]
@XComp I've pushed my changes, even though I know it will fail on the Python part, but wanted to see if everything else passes/is now OK.

**Comment by XComp** [comment]
Did you do a rebase onto `master` that includes the FLINK-32056 change? Because the python module is still failing for some reason :thinking:

**Comment by MartijnVisser** [comment]
> Did you do a rebase onto `master` that includes the [FLINK-32056](https://issues.apache.org/jira/browse/FLINK-32056) change? Because the python module is still failing for some reason 🤔

I did. I'm suspecting that the Shaded actually breaks something for Python as well, so looking into that.

**Comment by MartijnVisser** [comment]
> Because the python module is still failing for some reason 🤔

Ah, `flink-connector-jdbc` also uses `flink-shaded` :(

**Comment by XComp** [comment]
@MartijnVisser What's the state of this PR? I guess, we shouldn't wait for it to be merged for too long.

**Comment by MartijnVisser** [comment]
> @MartijnVisser What's the state of this PR? I guess, we shouldn't wait for it to be merged for too long.

There's a blocker with the JDBC connector also relying on `flink-shaded`. That connector vote first needs to be completed, before this is unblocked

**Comment by MartijnVisser** [comment]
@flinkbot run azure

**Comment by MartijnVisser** [comment]
@flinkbot run azure

File: flink-end-to-end-tests/flink-confluent-schema-registry/pom.xml [hunk 3]
```diff
@@ -53,6 +53,11 @@ under the License.
 			<artifactId>flink-connector-kafka</artifactId>
 			<version>3.0.0-1.17</version>
 		</dependency>
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-shaded-guava</artifactId>
+			<version>30.1.1-jre-16.1</version>
```
Comment (snuyanzin) [inline flink-end-to-end-tests/flink-confluent-schema-registry/pom.xml:59]: shouldn't it be 31.xx ?

Comment (snuyanzin) [reply flink-end-to-end-tests/flink-confluent-schema-registry/pom.xml:59]: yes, i see...
then would it make sense to have a comment that this version should be in sync with kafka-connector's?

File: pom.xml [hunk 4]
```diff
@@ -353,13 +353,13 @@ under the License.
 			<dependency>
 				<groupId>org.apache.flink</groupId>
 				<artifactId>flink-shaded-netty</artifactId>
-				<version>4.1.82.Final-${flink.shaded.version}</version>
+				<version>4.1.91.Final-${flink.shaded.version}</version>
```
Comment (XComp) [inline pom.xml:356]: We initially thought of doing the flink-shaded update early in the release to have enough CI cycles to test the functionality. With the delays we're getting close to the 1.18 feature freeze which contradicts the initial idea of the timing. I discussed this with @zentol: Based on past experience, netty was the source of instabilities. So, if we want to play it safe we could upgrade all the shaded dependencies except for netty and do the netty upgrade in 1.19 early on.

But on the other note: With 2 weeks until the feature freeze (which might be postponed anyway) and additional time for bug fixing and release testing, we still might have enough CI cycles. What's your opinion on that one?

I don't see a reason to not do the upgrade more conservatively (i.e. waiting with the netty upgrade and doing it in a separate ticket). Chesnay things that there is still enough time.

Comment (MartijnVisser) [reply pom.xml:356]: > With 2 weeks until the feature freeze (which might be postponed anyway) and additional time for bug fixing and release testing, we still might have enough CI cycles. What's your opinion on that one?

I think there's still enough time to actually merge the changes as-is.

File: flink-end-to-end-tests/flink-confluent-schema-registry/pom.xml (hunk 3 above)
Comment (MartijnVisser) [reply flink-end-to-end-tests/flink-confluent-schema-registry/pom.xml:59]: I'll add a comment when merging :)

**Review by XComp — APPROVED** [review]
LGTM :+1: 

> I'll add a comment when merging :)

Generally, I'm not a big fan of adding things "while merging" because it doesn't show up in the PR which makes it harder investigate issues on `master`. I'm still going to approve it because it's just a comment that shall be added. I guess, your motivation is to not have another CI run? :thinking:

File: pom.xml (hunk 4 above)
Comment (XComp) [reply pom.xml:356]: Ok, fine with me - I just wanted to bring this discussion up once. :+1:

**Comment by MartijnVisser** [comment]
> Generally, I'm not a big fan of adding things "while merging" because it doesn't show up in the PR which makes it harder investigate issues on `master`

Oh I always add it to the PR, but I don't wait for that CI cycle to fix (only if it's a comment)

**Review by snuyanzin — APPROVED** [review]

**Pull request merged**